Introduce support for the new "i4i.metal" AWS host instance type

dimitarproynov opened the pull request:
Use the VMC SDK 0.10.0, where the HostInstanceType enum has been replaced with string constants. Introduced a
method, that converts the host_instance_type format defined in the examples to the string constants coming from
the SDK. It's better to leave the host_instance_type format defined in the examples for backward compatibility with
the existing terraform managed SDDCs/Clusters.

Extracted the creation of schema for SDDC and Cluster into methods, so they can be used to mock schema.ResourceData
in the unit-tests.

Additionally, removed references to the R5 host instance type from the documentation, due to a shortage of these
hosts in AWS, see: https://github.com/vmware/terraform-provider-vmc/issues/128
The code that concerns them is left to enable users to manage their already deployed instances

Testing done:
Successfully deployed a single-host i4i.metal SDDC
Successfully destroyed a single-host i4i.metal SDDC

Successfully deployed a single-host i3i.metal SDDC
Successfully destroyed a single-host i3i.metal SDDC

golangci-lint run - passed
make test - passed

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

Commit:
Introduce support for the new "i4i.metal" AWS host instance type

Use the VMC SDK 0.10.0, where the HostInstanceType enum has been replaced with string constants. Introduced a
method, that converts the host_instance_type format defined in the examples to the string constants coming from
the SDK. It's better to leave the host_instance_type format defined in the examples for backward compatibility with
the existing terraform managed SDDCs/Clusters.

Extracted the creation of schema for SDDC and Cluster into methods, so they can be used to mock schema.ResourceData
in the unit-tests.

Additionally, removed references to the R5 host instance type from the documentation, due to a shortage of these
hosts in AWS, see: https://github.com/vmware/terraform-provider-vmc/issues/128
The code that concerns them is left to enable users to manage their already deployed instances

Testing done:
Successfully deployed a single-host i4i.metal SDDC
Successfully destroyed a single-host i4i.metal SDDC

Successfully deployed a single-host i3i.metal SDDC
Successfully destroyed a single-host i3i.metal SDDC

golangci-lint run - passed
make test - passed

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

fstrudel commented on vmc/resource_vmc_cluster_test.go:
```diff
@@ -172,3 +173,22 @@ func testAccVmcClusterResourceImportStateIdFunc(resourceName string) resource.Im
+
+	bareBonesClusterConfig, _ := buildClusterConfig(mockResourceSchema)
+
+	if *bareBonesClusterConfig.HostInstanceType != model.SddcConfig_HOST_INSTANCE_TYPE_I4I_METAL {
```
Can we have a table test to test all instance types, not just HostInstancetypeI4I?

+ We might also want to introduce some test assertions instead (here and other tests)? (easier to read I think -  like https://pkg.go.dev/github.com/stretchr/testify/assert)

dimitarproynov replied:
Great many thanks! I've been looking for TestNG/JUnit's assertEquals  equivalent for GO, but some unit tests at the aws terraform provider were using this lame equality test.

fstrudel commented on vmc/utils.go:
```diff
@@ -98,3 +98,20 @@ func getTotalSddcHosts(sddc *model.Sddc) int {
+	case HostInstancetypeR5:
+		return model.SddcConfig_HOST_INSTANCE_TYPE_R5_METAL, nil
+	default:
+		return "", fmt.Errorf("unknown host instance type")
```
You should add the given instance type in the error message.

fstrudel commented on vmc/utils_test.go:
```diff
@@ -0,0 +1,52 @@
+	"testing"
+)
+
+func TestToHostInstanceType(t *testing.T) {
```
Use a table driven test instead. You can also run in parallel: see https://dave.cheney.net/2019/05/07/prefer-table-driven-tests and https://www.gopherguides.com/articles/table-driven-testing-in-parallel

Commit:
Introduce support for the new "i4i.metal" AWS host instance type

Use the VMC SDK 0.10.0, where the HostInstanceType enum has been replaced with string constants. Introduced a
method, that converts the host_instance_type format defined in the examples to the string constants coming from
the SDK. It's better to leave the host_instance_type format defined in the examples for backward compatibility with
the existing terraform managed SDDCs/Clusters.

Extracted the creation of schema for SDDC and Cluster into methods, so they can be used to mock schema.ResourceData
in the unit-tests.

Additionally, removed references to the R5 host instance type from the documentation, due to a shortage of these
hosts in AWS, see: https://github.com/vmware/terraform-provider-vmc/issues/128
The code that concerns them is left to enable users to manage their already deployed instances

Testing done:
Successfully deployed a single-host i4i.metal SDDC
Successfully destroyed a single-host i4i.metal SDDC

Successfully deployed a single-host i3i.metal SDDC
Successfully destroyed a single-host i3i.metal SDDC

golangci-lint run - passed
make test - passed

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

Commit:
Convert unit tests to table driven convention

Make comparisons in the refactored unit test with the "assert" package

Testing done:
Successfully deployed a single-host i4i.metal SDDC
Successfully destroyed a single-host i4i.metal SDDC

Successfully deployed a single-host i3i.metal SDDC
Successfully destroyed a single-host i3i.metal SDDC

golangci-lint run - passed
make test - passed

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

fstrudel commented on vmc/resource_vmc_cluster_test.go:
```diff
@@ -172,3 +174,48 @@ func testAccVmcClusterResourceImportStateIdFunc(resourceName string) resource.Im
+
+	for _, testCase := range tests {
+		var mockResourceSchema = schema.TestResourceDataRaw(t, clusterSchema(), testCase.input)
+		got, _ := buildClusterConfig(mockResourceSchema)
```
You should not discard the error.
You should:
1. add a testcase that returns the error and check you got the correct one (so add an `expectedErr` field to your test struct
2. in all the other testcases, check that the returned error is `nil`


Also if the only thing that changes between tests is the instance type, write the tests to make that clear, like (untested):
```
	tests := []test{
		{instanceType: HostInstancetypeI3,
			expectedError: nil},
		{instanceType: HostInstancetypeI3EN,
			expectedError: nil},
		{instanceType: HostInstancetypeR5,
			expectedError: nil},
	}

	for _, testCase := range tests {
		config := map[string]interface{}{
			"num_hosts":          10,
			"host_instance_type": testCase.instanceType,
		}
                schema := schema.TestResourceDataRaw(t, clusterSchema(), config) // don't call it mock, there's no mock involved, you're calling all the real methods

		got, err := buildClusterConfig(&schema)
                assert.Equal(t, err, testCase.expectedError)
		if testCase.expectedError != nil {
                     // check error message
                }
	        assert.Equal(t, *got.HostInstanceType, testCase.instanceType)
	}
```

fstrudel commented on vmc/resource_vmc_sddc_test.go:
```diff
@@ -156,3 +158,95 @@ resource "vmc_sddc" "sddc_1" {
+				NumHosts:         MinHosts,
+				HostInstanceType: String(model.SddcConfig_HOST_INSTANCE_TYPE_I3_METAL),
+				Size:             String(MediumSDDCSize),
+				DeploymentType:   String(SingleAvailabilityZone),
```
Same here, don't repeat the whole struct here both for `input` and for `expected`. You can construct the input and expected structs in your testcase, just put what is different in this test compared to the next tests.

fstrudel commented on vmc/utils_test.go:
```diff
@@ -0,0 +1,36 @@
+	for _, testCase := range tests {
+		got, err := toHostInstanceType(testCase.input)
+		assert.Equal(t, got, testCase.want.converted)
+		assert.Equal(t, err, testCase.want.err)
```
If there was an error expected, you should check that the error message is the correct one.

dimitarproynov replied:
Internaly assert.Equal uses reflect.DeepEqual for non-primitive types, so it checks the error string inside the "error" object. I've tried changing a single char in the expected error message and the test failed.

fstrudel reviewed (approved):
You should check that you have 100% coverage for your new code (error s paths MUST be covered)

Commit:
Convert unit tests to table driven convention

Make comparisons in the refactored unit test with the "assert" package.

Testing done:
Successfully deployed a single-host i4i.metal SDDC
Successfully destroyed a single-host i4i.metal SDDC

Successfully deployed a single-host i3i.metal SDDC
Successfully destroyed a single-host i3i.metal SDDC

golangci-lint run - passed
make test - passed

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>
